### PR TITLE
Allow enums to be Iterable

### PR DESCRIPTION
**Pull request by bradallred**
I noticed we do this a few times which requires a lot of manual casting. This adds a convenient method for iterating an enum that takes the burden of casting onto itself and does so safely/efficiently by using the underlying type.

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] I used the same coding style as the surrounding code 
- [x] I have tested the proposed changes
- [x] I extended the documentation, if necessary
- [x] The proposed change builds also on our build bots (check after submission)

### CODE REVIEW

**Comment by lynxlynxlynx** [comment]
Now you can remove the new begin/end members in GUIColors and HCStrings.

**Comment by bradallred** [comment]
I did think about it, but that lends itself to non obvious bugs when somebody wants to add something to them. It's also a bit more ergonomic to do `for(const EnumType e : EnumIterator<EnumType>())` as opposed to `for(const EnumType e : EnumIterator<EnumType, Begin, End>())`.

**Comment by lynxlynxlynx** [comment]
Yeah, fetching the first element ourselves would already be better. There are not many users and half of them already pass all the data.

File: gemrb/core/WorldMap.h [hunk 1]
```diff
@@ -94,8 +96,8 @@ class GEM_EXPORT WMPAreaEntry {
 	ieStrRef LocCaptionName = ieStrRef::INVALID;
 	ieStrRef LocTooltipName = ieStrRef::INVALID;
 	ResRef LoadScreenResRef;
-	ieDword AreaLinksIndex[4]{};
-	ieDword AreaLinksCount[4]{};
+	EnumArray<WMPDirection, int> AreaLinksIndex;
+	EnumArray<WMPDirection, int> AreaLinksCount;
```
Comment (lynxlynxlynx) [inline gemrb/core/WorldMap.h:100]: Why the change of type, doesn't it just create more work for us in the importer?

File: gemrb/core/EnumIndex.h [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	static_assert(std::is_unsigned<under_t<ENUM>>::value, "EnumIndex must be unsigned");
+
+	under_t<ENUM> val;
+public:
+	explicit EnumIterator(const ENUM& e) : val(UnderType(e)) {}
+	EnumIterator() : EnumIterator(BEGIN) {}
+	EnumIterator operator++() {
+		++val;
+		return *this;
+	}
```
Comment (lynxlynxlynx) [inline gemrb/core/EnumIndex.h:56]: Is this good enough? Specifically for sparse enums like `Direction` in FogRenderer.

Comment (bradallred) [reply gemrb/core/EnumIndex.h:56]: C++ doesn't have good enough reflection to constrain or increment any other way. We just have to know what we are doing.

We don't use `EnumIterator` with `FogRenderer::Direction`, but if we did we should iterate over "invalid" Directions such as `N|S`. The "invalid" directions would still exist inside an `EnumArray` tho, you can see this with `BAM_FLAGS` and we just fill those invalid gaps with `BlitFlags::NONE`. These slot are just unused filler.

File: gemrb/core/WorldMap.h (hunk 1 above)
Comment (bradallred) [reply gemrb/core/WorldMap.h:100]: a type change had to occur somewhere and the compiler complained about `GUIScript` first, which uses `int` :)

I don't know what is preferable, AFICT it doesn't matter which type we use (`int` vs `ieDword`).

Comment (lynxlynxlynx) [reply gemrb/core/WorldMap.h:100]: I'd prefer to keep matching the "spec", which uses ieDword.

File: gemrb/core/WorldMap.cpp [hunk 3]
```diff
@@ -444,9 +439,9 @@ void WorldMap::SetEncounterArea(const ResRef& area, const WMPAreaLink *link) {
 	AddAreaLink(std::move(ldest));
 	AddAreaLink(std::move(lsrc));
 
-	for (i = 0; i < 4; ++i) {
-		ae.AreaLinksCount[i] = 2;
-		ae.AreaLinksIndex[i] = static_cast<ieDword>(idx);
+	for (WMPDirection dir : EnumIterator<WMPDirection>()) {
+		ae.AreaLinksCount[dir] = 2;
+		ae.AreaLinksIndex[dir] = static_cast<int>(idx);
```
Comment (lynxlynxlynx) [inline gemrb/core/WorldMap.cpp:444]: One more here.
